src/tests/suite/daos_container.c

```diff
@@ -771,6 +771,226 @@ co_set_prop(void **state)
+	assert_non_null(cont_acl);
+	rc = daos_acl_get_ace_for_principal(cont_acl, DAOS_ACL_OWNER, NULL,
+					    &update_ace);
+	assert_int_equal(rc, 0);
+	update_ace->dae_allow_perms = DAOS_ACL_PERM_CONT_ALL;
+
+	print_message("- getting container handle\n");
+	rc = daos_cont_open(arg->pool.poh, arg->co_uuid, DAOS_COO_RW, &coh,
+			    NULL, NULL);
+	assert_int_equal(rc, 0);
+
+	print_message("- updating cont ACL to restore delete privs\n");
+	rc = daos_cont_update_acl(coh, cont_acl, NULL);
+	assert_int_equal(rc, 0);
+
+	print_message("- closing container\n");
+	rc = daos_cont_close(coh, NULL);
+	assert_int_equal(rc, 0);
+
+	print_message("Deleting container now should succeed\n");
+	rc = daos_cont_destroy(arg->pool.poh, arg->co_uuid, 1, NULL);
+	assert_int_equal(rc, 0);
+
+	MPI_Barrier(MPI_COMM_WORLD);
+
+	/* Clear cont uuid since we already deleted it */
+	uuid_clear(arg->co_uuid);
+
+	daos_acl_free(cont_acl);
+	daos_prop_free(pool_prop);
+	daos_prop_free(cont_prop);
+	test_teardown((void **)&arg);
+}
+
+static void
+co_destroy_allowed_by_pool(void **state)
+{
+	test_arg_t	*arg0 = *state;
+	test_arg_t	*arg = NULL;
+	daos_prop_t	*pool_prop;
+	daos_prop_t	*cont_prop;
+	int		 rc;
+
+	rc = test_setup((void **)&arg, SETUP_EQ, arg0->multi_rank,
+			DEFAULT_POOL_SIZE, NULL);
+	assert_int_equal(rc, 0);
+
+	print_message("Delete a container where only pool allows it\n");
+
+	/* pool gives the owner all privs, including delete cont */
+	pool_prop = get_daos_prop_with_owner_acl_perms(DAOS_ACL_PERM_POOL_ALL,
+						       DAOS_PROP_PO_ACL);
+
+	/* container doesn't give delete privs to the owner */
+	cont_prop = get_daos_prop_with_owner_acl_perms(DAOS_ACL_PERM_CONT_ALL &
+						       ~DAOS_ACL_PERM_DEL_CONT,
+						       DAOS_PROP_CO_ACL);
+
+	while (!rc && arg->setup_state != SETUP_CONT_CREATE)
+		rc = test_setup_next_step((void **)&arg, NULL, pool_prop,
+					  cont_prop);
+	assert_int_equal(rc, 0);
+
+	MPI_Barrier(MPI_COMM_WORLD);
+
+	print_message("Deleting container should succeed\n");
+	rc = daos_cont_destroy(arg->pool.poh, arg->co_uuid, 1, NULL);
+	assert_int_equal(rc, 0);
+
+	MPI_Barrier(MPI_COMM_WORLD);
+
+	/* Clear cont uuid since we already deleted it */
+	uuid_clear(arg->co_uuid);
+
+	daos_prop_free(pool_prop);
+	daos_prop_free(cont_prop);
+	test_teardown((void **)&arg);
+}
+
+static void
+expect_cont_open_access(test_arg_t *arg0, uint64_t perms, uint64_t flags,
+			int exp_result)
+{
+	test_arg_t	*arg = NULL;
+	daos_prop_t	*prop;
+	int		 rc;
+
+	rc = test_setup((void **)&arg, SETUP_EQ, arg0->multi_rank,
+			DEFAULT_POOL_SIZE, NULL);
+	assert_int_equal(rc, 0);
+
+	prop = get_daos_prop_with_owner_acl_perms(perms,
+						  DAOS_PROP_CO_ACL);
+
+	while (!rc && arg->setup_state != SETUP_CONT_CREATE)
+		rc = test_setup_next_step((void **)&arg, NULL, NULL, prop);
+	assert_int_equal(rc, 0);
+
+	MPI_Barrier(MPI_COMM_WORLD);
+
+	rc = daos_cont_open(arg->pool.poh, arg->co_uuid, flags, &arg->coh,
```

Comment:
may want to have just rank 0 do this. If that is in fact the right thing, then I wonder if test_setup_cont_open() could be made more flexible (to accept the flags we want to pass to the API) - so the pattern of run on rank 0 and bcast the rc to all other ranks could be captured in a single function call. Might get a little messy trying to invoke it directly rather than how it normally is invoked through that test_setup_next_step() regime.

Reply:
After looking at this, I'm going to skip all of these tests if not running on rank 0. It seems like in practice the test setup always does the create/destroy/opens only on rank 0, and the logic I'm trying to test shouldn't vary based on rank anyway.

I updated the test_arg_t struct to accept the flags for both pool connect and container open, so I can just use the existing test infrastructure.